utils/forest_utils/src/lib.rs

```diff
@@ -1,14 +1,135 @@
 // Copyright 2019-2023 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-pub extern crate const_format;
-
 pub mod db;
 pub mod encoding;
 pub mod io;
 pub mod json;
-pub mod macros;
 pub mod misc;
 pub mod monitoring;
 pub mod net;
 pub mod version;
+
+use std::{pin::Pin, time::Duration};
+
+use futures::{
+    future::{pending, FusedFuture},
+    select, Future, FutureExt,
+};
+use tokio::time::sleep;
+use tracing::error;
+
+/// Keep running the future created by `make_fut` until the timeout or retry
+/// limit in `args` is reached.
+/// `F` _must_ be cancel safe.
+// TODO(aatifsyed): store most recent error in RetryError
+// REFACTOR?(aatifsyed): use `futures-retry` or `again`
+// REFACTOR?(aatifsyed): return num retries
+#[tracing::instrument(skip(make_fut))]
+pub async fn retry<F, T, E>(
+    args: RetryArgs,
+    mut make_fut: impl FnMut() -> F,
+) -> Result<T, RetryError>
+where
+    F: Future<Output = Result<T, E>>,
+    E: std::fmt::Debug,
+{
+    let mut timeout: Pin<Box<dyn FusedFuture<Output = ()>>> = match args.timeout {
+        Some(duration) => Box::pin(sleep(duration).fuse()),
+        None => Box::pin(pending()),
+    };
+    let max_retries = args.max_retries.unwrap_or(usize::MAX);
+    let mut task = Box::pin(
+        async {
+            for _ in 0..max_retries {
+                match make_fut().await {
+                    Ok(ok) => return Ok(ok),
+                    Err(err) => error!("retrying operation after {err:?}"),
+                }
+                if let Some(delay) = args.delay {
+                    sleep(delay).await;
+                }
+            }
+            Err(RetryError::RetriesExceeded)
+        }
+        .fuse(),
+    );
+    select! {
+        _ = timeout => Err(RetryError::TimeoutExceeded),
+        res = task => res,
+    }
+}
+
+#[derive(Debug, Clone, Copy, smart_default::SmartDefault)]
+pub struct RetryArgs {
+    #[default(Some(Duration::from_secs(1)))]
+    pub timeout: Option<Duration>,
+    #[default(Some(5))]
+    pub max_retries: Option<usize>,
+    #[default(Some(Duration::from_millis(200)))]
+    pub delay: Option<Duration>,
+}
+
+impl RetryArgs {
+    pub fn new_ms(
```

Comment:
Does it simplify the client code a lot? Otherwise, I'd keep it general and accept any `Duration`. It would also allow smaller values for tests.

Reply:
It does simplify code - I've moved this to the test module for now, which is where it was used.